lapack/gonum/dgesc2.go

```diff
@@ -0,0 +1,57 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+// Dgesc2 solves a system of linear equations
+//   A * X = scale* RHS
+// with a general N-by-N matrix A using the LU factorization with
+// complete pivoting computed by Dgetc2.
+func (impl Implementation) Dgesc2(n int, a []float64, lda int, rhs []float64, ipiv, jpiv []int) (scale float64) {
+	const smlnum = dlamchS / dlamchP
+	if len(a) < (n-1)*lda+n {
+		panic(shortA)
+	}
+
+	// Apply permutations ipiv to RHS.
+	impl.Dlaswp(1, rhs, lda, 1, n-1, ipiv, 1)
```

Comment:
I thought it's strange that `k2` is `n-1` and not `n` in the Fortran reference. In fact it's because they rely on an undocumented implementation detail of DGETC2:
```
*
*     Set last pivots to N
*
      IPIV( N ) = N
      JPIV( N ) = N
```
I think it would be cleaner to pass `0, n-1, ipiv[:n], 1` here. What do you think, @kortschak ?

Reply:
Wouldn't that mean were doing an additional row of work that is in effect a no-op?

Reply:
Yes but only because our `Dlaswp` implementation doesn't check whether it's doing a no-op swap (unlike the reference which does the check). I think it would be better to pass `n` pivots here and below and change `Dlaswp` to check for a no-op.

Reply:
Yeah, OK.

Reply:
#1654 submitted